CK2FileManager.delegateQueue

mikeabdullah opened the pull request:
Like `NSURLSession`, `CK2FileManager` now makes it explicit which queue delegate methods and blocks are executed on. This makes the implementation of things like the open panel somewhat simpler. Hopefully you're OK with this @MrNoodle ?

mikeabdullah committed:
Give file manager a delegate queue
Not actually used yet though!

mikeabdullah committed:
Send delegate messages on delegate queue

mikeabdullah committed:
Execute completion handler on delegate queue

mikeabdullah committed:
Execute enumeration block on delegate queue

mikeabdullah committed:
Execute progress block on delegate queue

mikeabdullah committed:
Let file manager handle main thread serialisation for CKUploader

mikeabdullah committed:
Take advantage of scheduling open panel's callbacks on main queue

mikeabdullah committed:
Remove NSObject+CK2OpenPanel
It's no longer needed as file manager takes care of bouncing over to main queue

MrNoodle commented:
I think there is a problem with getting rid of the NSObject category (at least without fixing the code that calls it). If you look at the comment it points out an issue with using any sort of dispatch queue. The block is supplied by the client and can potentially call something like -runModal:. That method blocks while still allowing the panel to be operated which may queue up more stuff to be executed on the queue. As a result, any execution queues (which I think would also affect operation queues), would end up deadlocked. The method on NSObject specifically calls -performSelectorOnMainThread: which is on a separate system altogether and gets around this issue.

mikeabdullah commented:
Ah, think I see what you mean now. I can see two scenarios to worry about:

### 1. Client calls -runModal from main queue:

```
[[NSOperationQueue mainQueue] addOperationWithBlock:^{
    [openPanel runModal];
}];
```

Clearly that's rather contrived, but the important point is the open panel is now blocking the main queue since that's where client chose to schedule it.

Looking around, it seems `NSOpenPanel` suffers in the same way: http://lists.apple.com/archives/cocoa-dev/2012/Dec/msg00279.html

Symptom in the UI is simply that the panel never finishes loading; it's stuck merrily spinning away. My inclination is to note this in the docs for `-runModal` (and perhaps the Readme too) and advise that people do not call it in a fashion which blocks the main queue. I feel `-runModal` tends to be the inferior presentation choice too, and asynchronous approaches are preferable when possible.

### 2. Client runs a modal panel from a callback

```
- (void)panel:(id)sender didReceiveAuthenticationChallenge:(NSURLAuthenticationChallenge *)challenge {
    [self presentError:challenge.error];
}
```

This will use a modal alert panel to present the error. While the alert is visible, the main queue is blocked so no more delegate methods or block callbacks can be delivered.

I don't consider this too big of a deal though. The panel blocks access to the rest of the UI, so it's not like the user could interact with it anyway. All they're potentially missing out on is some feedback in the background as the panel loads in some data.

Besides, clients can avoid this too if they wish by punting their modal work until the next iteration of the runloop using `-performSelector:afterDelay:…`

**Update:** ah, looks like something along these lines causes a problem for the New Folder panel. Am investigating

mikeabdullah committed:
Merge branch 'v2.x-beta' into delegate-queue

mikeabdullah committed:
Remove old #import

MrNoodle commented:
A less contrived example is if someone does -runModal in the completion handler for -setDirectory. This is where I ran into it as I did not want to present the open panel until the directory change had completed (so that I didn't have to show a in-progress open panel).

mikeabdullah commented:
Ah, I did wonder if you might be doing something like that. What do you do with your UI in the gap while the directory change occurs? With some connections/servers this could be pretty long.

MrNoodle commented:
I have a progress spinner in the previous panel where you set all the server parameters to connect to.

MrNoodle commented:
As far as having the client call -perform..., I feel like if we can easily avoid this (which I feel we already did), then we should. The fix was already there so I prefer to stick with it to make things easier for the client.

mikeabdullah committed:
Merge branch 'v2.x-beta' into delegate-queue

Conflicts:
	ConnectionKit/CK2FileOperation.m
	ConnectionKit/CK2OpenPanelController.m
	ConnectionKit/CKUploader.m

mikeabdullah committed:
Let delegate queue bounce auth and transcript messages to main thread for us too

mikeabdullah committed:
Actually bounce blocks over to delegate queue

mikeabdullah committed:
Merge branch 'v2.x-beta' into delegate-queue

Conflicts:
	ConnectionKit/CK2FileOperation.m
	ConnectionKit/CK2OpenPanelController.m

mikeabdullah committed:
Complain when trying ot message non-existent file manager

mikeabdullah committed:
Don't try to ditch file manager reference early

mikeabdullah committed:
Merge branch 'v2.x-beta' into delegate-queue

Conflicts:
	ConnectionKit/CK2FileOperation.m
	ConnectionKit/CKUploader.m

mikeabdullah committed:
Merge branch 'v2.x-beta' into delegate-queue

Conflicts:
	ConnectionKit/CKUploader.m
	ConnectionKit/NSObject+CK2OpenPanel.m

mikeabdullah commented:
I've done a little more testing for case 1 as laid out above. Aside from our own callback delivery, it turns out this is also a big no-no on the part of clients since bouncy scrolling is broken under these circumstances too. I'm adding a note to `-runModal` about that.

mikeabdullah committed:
Rip off NSSavePanel's docs

mikeabdullah committed:
Cautionary note about -runModal from within the main dispatch queue

mikeabdullah committed:
Specially execute -setDirectoryURL:completionHandler: completion handler outside of a dispatch queue

mikeabdullah commented:
So I think I've reached a pretty decent compromise where the open panel internally allows its file manager to use the `+mainQueue` but then handles `-setDirectoryURL:completionHandler:` specially so as to make clients' life easier. How do you feel about it now @MrNoodle?

mikeabdullah committed:
Merge branch 'v2.x-beta' into delegate-queue

Conflicts:
	ConnectionKit/CK2FileOperation.m

mikeabdullah committed:
Remove old note about threading

mikeabdullah commented:
Think I'm going to merge this in in a few days, assuming there's no complaints

mikeabdullah committed:
Class method for creating file managers instead
Might as well match NSURLSession here; gives us a little more future flexibility

mikeabdullah commented on ConnectionKit/CK2OpenPanelController.m:
```diff
@@ -523,20 +517,12 @@ - (void)changeDirectory:(NSURL *)directoryURL completionHandler:(void (^)(NSErro
                 [self validateViews];
                 
                 [resolvedURL release];
```
@MrNoodle when I run the static analyser on this branch, it complains that the release of `resolvedURL` here is unwarranted. Looking at the code I agree with it. It looks to me like the fact this was wrapped in a `dispatch_async` block before was effectively hiding it from the analyser.

Is the analyser wrong here, or our code?

mikeabdullah replied:
@MrNoodle I've made commit b9a2eed08c3fc91dfcf801f10dacf6cdeaba369f over on the main 2.x branch to fix this. Can back it out if it turns out to be a mistake.

MrNoodle replied:
That's fine. I think it was a remnant from an earlier version of that code. Also, you can probably get rid of the unnecessary retain/release pair at lines 506 & 525 since the block should be retaining resolvedURL in that case.

mikeabdullah replied:
Ta, I've done that in 66dac9c50d342c7164d54276b6b9ca58b4c58383

mikeabdullah committed:
Retain file manager, not over-release it!

mikeabdullah committed:
Message delegate on same queue as file manager

mikeabdullah committed:
Merge branch 'v2.x-beta' into delegate-queue

Conflicts:
	ConnectionKit/CK2OpenPanelController.m

mikeabdullah committed:
Merge branch 'v2.x-beta' of https://github.com/karelia/ConnectionKit into delegate-queue

Conflicts:
	ConnectionKit/CK2OpenPanelController.m
